Python/Include/Sample.hpp

```diff
@@ -0,0 +1,48 @@
+#ifndef ROVER_PYTHON_SAMPLE_HPP
+#define ROVER_PYTHON_SAMPLE_HPP
+#include <string_view>
+#include <pybind11/pybind11.h>
+
+namespace Rover {
+
+  //! Sample containing Python objects.
+  struct PythonSample {
+    using Result = pybind11::object;
+    using Parameters = pybind11::tuple;
+
+    //! The result of the function.
+    Result m_result;
+
+    //! The arguments passed to the function.
+    Parameters m_arguments;
+  };
+
+  //! Exports a Sample for Python objects.
+  /*!
+    \param module The module to export the class to.
+  */
+  void export_sample(pybind11::module& module);
+
+  //! Exports a Sample for specified types.
+  /*!
+    \tparam R The type of the result.
+    \tparam P The types of the parameters.
+    \param module The module to export the class to.
+    \param postifx The postfix to the name of the exported class.
+  */
+  template<typename R, typename... P>
+  void export_sample(pybind11::module& module, std::string_view postfix);
+
+  template<typename R, typename... P>
+  void export_sample(pybind11::module& module, std::string_view postfix) {
+    auto name = std::string("Constant").append(postfix);
+    pybind11::class_<Sample<R, P...>>(module, name.c_str())
+      .def(pybind11::init<>())
+      .def_readwrite("result", &Sample<R, P...>::m_result)
+      .def_readwrite("arguments", &Sample<R, P...>::m_arguments);
+    pybind11::implicitly_convertible<Sample<R, P...>, PythonSample>();
```

Comment:
You will need to actually implement such a conversion rather than only declare that one exists.

Reply:
PythonSample -> Sample<R, P...> conversion is relatively straightforward as I get to define a special constructor for Sample<R, P...>. The opposite is, however, not so much.
One way is to explore the type_caster thing under pybind11::detail further but it feels like patchwork. Another approach is to expose PyObject instantiations of pybind11::class_<T> - that is probably better than it sounds as generic exports implicitly depend on PyObject exports anyway, and the change will make this relationship explicit, hence imposing the correct order of exports. What do you think?